ledger/financial/cmd/client/batcher.go

```diff
@@ -0,0 +1,226 @@
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"ledger/financial/handler"
+	"ledger/financial/payload"
+	"ledger/financial/state"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/hyperledger/sawtooth-sdk-go/protobuf/batch_pb2"
+	"github.com/hyperledger/sawtooth-sdk-go/signing"
+	"google.golang.org/protobuf/proto"
+
+	tx_pb2 "github.com/hyperledger/sawtooth-sdk-go/protobuf/transaction_pb2"
+)
+
+// batcher defines the primitive to send transactions
+type batcher interface {
+	sendTransaction(payload payload.FinancialPayload) (string, error)
+}
+
+// newBatcher returns a new initialized batcher. url must be the REST endpoint,
+// and key must by the filepath to a private key that is allowed to send
+// batches.
+func newBatcher(url url.URL, keyfilePath string) (batcher, error) {
+	privateKey, err := loadKey(keyfilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load key: %v", err)
+	}
+
+	pk := signing.NewSecp256k1Context().GetPublicKey(privateKey)
+	fmt.Println("pk:", pk.AsHex())
+
+	cryptoFac := signing.NewCryptoFactory(signing.NewSecp256k1Context())
+	signer := *cryptoFac.NewSigner(privateKey)
+
+	return basicBatcher{
+		url:    url,
+		signer: signer,
+	}, nil
+}
+
+// batcher holds the functionality of creating and sending batches
+type basicBatcher struct {
+	url    url.URL
+	signer signing.Signer
+}
+
+// sendTransaction implements batcher
+func (b basicBatcher) sendTransaction(payload payload.FinancialPayload) (string, error) {
+
+	payloadBuff, err := payload.ToBytes()
+	if err != nil {
+		return "", fmt.Errorf("failed to transform payload into bytes: %v", err)
+	}
+
+	hash := sha512.New()
+	hash.Write(payloadBuff)
+
+	payloadSha512 := strings.ToLower(hex.EncodeToString(hash.Sum(nil)))
+
+	address := state.KeyAddress
+
+	nonce := make([]byte, 16)
+
+	_, err = rand.Read(nonce)
+	if err != nil {
+		return "", fmt.Errorf("failed to get random nonce: %v", err)
+	}
+
+	rawTxHeader := tx_pb2.TransactionHeader{
+		SignerPublicKey:  b.signer.GetPublicKey().AsHex(),
+		FamilyName:       handler.FamilyName,
+		FamilyVersion:    handler.FamilyVersion,
+		Dependencies:     []string{},
+		Nonce:            base64.RawStdEncoding.EncodeToString(nonce),
+		BatcherPublicKey: b.signer.GetPublicKey().AsHex(),
+		Inputs:           []string{address},
+		Outputs:          []string{address},
+		PayloadSha512:    payloadSha512,
+	}
+
+	txHeader, err := proto.Marshal(&rawTxHeader)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal transaction header: %v", err)
+	}
+
+	txHeaderSig := hex.EncodeToString(b.signer.Sign(txHeader))
+
+	transaction := tx_pb2.Transaction{
+		Header:          txHeader,
+		HeaderSignature: txHeaderSig,
+		Payload:         payloadBuff,
+	}
+
+	rawBatchList, err := b.createBatch([]*tx_pb2.Transaction{&transaction})
+	if err != nil {
+		return "", fmt.Errorf("failed to create batch: %v", err)
+	}
+
+	// batchID := rawBatchList.Batches[0].HeaderSignature
```

Comment:
left over commented out line ? maybe add TODO ?

Reply:
This will be needed to get the batch receipt and check the result.